Add Confirmation Before Dev Tool Removal and Modification [`AARD-2033`]

ryanzhangofficial opened the pull request:
## Task
Add a confirmation dialog when removing a Dev Tools–placed scoring zone, allowing the user to either remove it temporarily (current behavior) or permanently by also deleting it from the Dev Tools cache.

[Jira Ticket](https://jira.autodesk.com/browse/AARD-2033)

## Symptom
Currently if you remove a protected zone or scoring zone from the configure field panel that is placed by the dev tool it will remove it until the next refresh. Only the dev tool is able to cache scoring zone placement.

## Solution
Popup a confirmation window when we remove a scoring zone that was placed by dev tools, and give us the option of removing it from dev tools as well (saving to cache).

## Verification
- Tested temporary removal + refresh
- Tested permanent removal + refresh

<img width="3839" height="2043" alt="Screenshot 2025-08-08 102909" src="https://github.com/user-attachments/assets/dfc4d33b-9af5-4562-9ec2-870992edd5a6" />

---

Before merging, ensure the following criteria are met:

- [ ] All acceptance criteria outlined in the ticket are met.
- [ ] Necessary test cases have been added and updated.
- [ ] A feature toggle or safe disable path has been added (if applicable).
- [ ] User-facing polish:
  - Ask: *"Is this ready-looking?"*
- [ ] Cross-linking between Jira and GitHub:
  - PR links to the relevant Jira issue.
  - Jira ticket has a comment referencing this PR.

ryanzhangofficial committed:
Add zone utils

ryanzhangofficial committed:
Add zone removal modal

ryanzhangofficial committed:
chore: formatting

ryanzhangofficial committed:
Merge branch 'dev' into ryan/2033/add-confirmation-before-dev-tool-removal

rutmanz reviewed (changes requested):
I think it would be better to say that the zones were defined in the field file/asset file/mirabuf file and avoid all reference to the developer tools. I think that, in theory, the developer tools are used by us to add this information to the cached files that people download, and so we shouldn't expect that they've used them / know they're there.

Also, this doesn't handle modifying the zones

ryanzhangofficial committed:
Update toasts and descs

ryanzhangofficial committed:
Add permanent modification

ryanzhangofficial committed:
Update temporary modification

ryanzhangofficial committed:
chore: formatting

ryanzhangofficial committed:
Merge branch 'dev' into ryan/2033/add-confirmation-before-dev-tool-removal

ryanzhangofficial committed:
cleanup

rutmanz commented on fission/src/ui/modals/DevtoolZoneRemovalModal.tsx:
```diff
@@ -0,0 +1,110 @@
+    }
+
+    return (
+        <Dialog open={isOpen} onClose={onClose} maxWidth="sm" fullWidth>
```
I think it's fine to make language consistent rather than doing this ternary stuff. Removing a zone is a modification or change to the field, and I think it makes the code a lot cleaner to use the same words

rutmanz commented on fission/src/ui/modals/DevtoolZoneRemovalModal.tsx:
```diff
@@ -0,0 +1,110 @@
+                            <strong>Temporary {actionType === "modification" ? "modification" : "removal"}:</strong>{" "}
+                            {actionType === "modification"
+                                ? "Save changes until next field reload. Original zone will reappear when you refresh or reload the field."
+                                : "Remove zone until next field reload. The zone will reappear when you refresh or reload the field."}
```
"refresh or reload the field" is kind of a weird expression, "refresh the page" is much clearer IMO

rutmanz commented on fission/src/ui/modals/DevtoolZoneRemovalModal.tsx:
```diff
@@ -0,0 +1,110 @@
+                        <Typography variant="body2">
+                            <strong>Permanent {actionType === "modification" ? "modification" : "removal"}:</strong>{" "}
+                            {actionType === "modification"
+                                ? "Save changes to the field file cache. This will persist your modifications on future loads."
```
"Save changes to the local asset file. This will persist your modifications until you remove it from the cache."

rutmanz commented on fission/src/ui/modals/DevtoolZoneRemovalModal.tsx:
```diff
@@ -0,0 +1,110 @@
+    actionType?: "removal" | "modification"
+}
+
+const DevtoolZoneRemovalModal: React.FC<DevtoolZoneRemovalModalProps> = ({
```
should probably be renamed if this is being generalized

rutmanz commented on fission/src/util/DevtoolZoneUtils.ts:
```diff
@@ -0,0 +1,171 @@
+
+    const editor = new FieldMiraEditor(parts)
+
+    if (zoneType === "scoring") {
```
I don't think it makes very much sense to have all of this in this conditional, it looks like you can generalize this for either zone type and avoid making unnecessarily specific code

rutmanz commented on fission/src/util/DevtoolZoneUtils.ts:
```diff
@@ -0,0 +1,171 @@
+                devZone.name === originalZone.name &&
+                devZone.alliance === originalZone.alliance &&
+                devZone.parentNode === originalZone.parentNode &&
+                JSON.stringify(devZone.deltaTransformation) === JSON.stringify(originalZone.deltaTransformation)
```
it appears you're using these properties because they overlap on the two different zone preferences types. I think it would make sense to enforce this more rigorously by having a `BaseZonePreferences` type with the common properties that they both extend. This would be basically the same as `ScoringZonePreferences|ProtectedZonePreferences`, except that it would be robust aginst changes to those types

rutmanz replied:
Alternatively, generating IDs on load for the zone preferences objects (they don't need to persist in the cache or across reloads) might be a more robust way to compare two zones that doesn't require the structuredclone

ryanzhangofficial committed:
Update removal to modification

ryanzhangofficial committed:
Update DevtoolZoneUtils for restructure & BaseZonePreferences

ryanzhangofficial committed:
Update interfaces

ryanzhangofficial committed:
chore: formatting

rutmanz commented on fission/src/util/DevtoolZoneUtils.ts:
```diff
@@ -21,22 +33,14 @@ export function isZoneFromDevtools(
         return false
     }
+    
+    const devtoolZones = editor.getUserData("devtool:scoring_zones") as ScoringZonePreferences[] | undefined
```
You shouldn't need to cast this, if you do then you should update the interface to match instead

rutmanz reviewed (approved):
otherwise lgtm

AlexD717 commented:
Why are we only adding this confirmation for scoring zones?


_Also Biome test fail_

ryanzhangofficial commented:
> Why are we only adding this confirmation for scoring zones?
> 
> _Also Biome test fail_

As noted in comments in code, protected zones are not supported yet.

AlexD717 commented:
> > Why are we only adding this confirmation for scoring zones?
> 
> As noted in comments in code, protected zones are not supported yet.

Yeah, I can see that, I am just wondering why we are only adding this for scoring zones. Are we planning to add this for protected zones later on? From a user perspective it feels weird if they are only getting the popup for scoring zones, and nothing else.

ryanzhangofficial commented:
> > > Why are we only adding this confirmation for scoring zones?
> > 
> > 
> > As noted in comments in code, protected zones are not supported yet.
> 
> Yeah, I can see that, I am just wondering why we are only adding this for scoring zones. Are we planning to add this for protected zones later on? From a user perspective it feels weird if they are only getting the popup for scoring zones, and nothing else.

Yes that's the plan.

I think the order of merging should be this PR -> Zach's protected zone pr -> protected zone removal and modification

ryanzhangofficial committed:
chore: formatting

AlexD717 commented:
Is it also intentional that you have to go into developer tools and press save to actually have scoring zones persist in cache? Currently spawning in a scoring zone and then refreshing doesn't have it persist. If this is the case, I am personally not a fan of this, and think that they should auto cache.


https://github.com/user-attachments/assets/62a81bcf-1ba9-4a75-86f7-dd0bb0804a42

ryanzhangofficial committed:
fix: save/cancel button dissappearing

rutmanz commented:
I agree with @AlexD717, but this appears to be the current behavior on dev as well

ryanzhangofficial commented:
> I agree with @AlexD717, but this appears to be the current behavior on dev as well

Yes I agree too, currently fixing this issue. It isn't a problem on prod so must've been an issue that came up throughout the summer.

ryanzhangofficial committed:
Add caching for manual scoring zones

ryanzhangofficial committed:
Update devtool panel to include devhandlers only

ryanzhangofficial commented:
Would appreciate thoughts on this approach. I thought it would be appropriate because of the rewording of the modals, but at the same time it makes devtools seem less useful.

BrandonPacewic committed:
fix: correct async handling, zone data safety, and cleanup in devtool zone flow

  - Await handleSave() before building modifiedZone in handlePermanentModification
    so deltaTransformation is read after the gizmo updates it
  - Remove internal try/catch from handlePermanentModification so errors propagate
    to the modal's catch block and the error toast actually fires on failure
  - Stop overwriting fieldPreferences.scoringZones with only devtool-cached zones
    in removeZoneFromDevtools, modifyZoneInDevtools, and addUserZoneToDevtools —
    previously dropped any user zones not yet in the cache
  - Split combined pause/save effect in ManageScoring/ProtectedZonesInterface so
    holdPause and releasePause only fire on mount/unmount, not on every zone change
  - Remove debug console.log that fired every second in DeveloperToolPanel
  - Remove placeholder devtool keys (a, b, test, keep, drop, bad, foo) from
    DevtoolMiraData interface; update tests to use real keys
  - Fix misplaced import statement in ConfigureScoringZonesInterface
  - Remove unnecessary IIFEs in zone row renders
  - Fix .ts extension on import in FieldMiraEditor

BrandonPacewic committed:
chore: merge `dev`

* dev: (297 commits)
  fix: remove broken dep
  fix: stabilize Black/Isort CI format validation
  fix: remove dead entries in `EventDataMap`
  fix: cleanups + prevent test listeners persist on window
  fix: don't add new zone to array before navigating to config
  fix: revert `fission/src/util/TypeConversions.ts` formatting
  fix: biome fmt issues
  fix: build isuses
  fix: correct zone refactor regressions
  fix: biome lint warnings
  feat: import unstable mira files with warnings
  revert: `2fd4c67aae0be43b473ed0e5b5f2d90f17ecfcf6`
  feat: handle zero triangles in `createConcaveShapeSettingsFromPart`
  doc: add current status to in development features
  fix: run formatter
  docs: update manual exporter install tutorial
  docs: tutorial content updates
  fix: spelling, grammer, and formatting corrections
  Update tutorials/CommandPalette.md
  Update tutorials/SpawnAsset.md
  ...

BrandonPacewic committed:
chore: add `tsconfig.tsbuildinfo` to `.gitignore`

BrandonPacewic committed:
fix: format issues

BrandonPacewic committed:
fix: biome type warnings/errors

BrandonPacewic committed:
fix: biome errors

BrandonPacewic committed:
fix: biome errors

BrandonPacewic committed:
fix: biome errors
